File: examples/go/internal/auth/usecase/jwt.go
package usecase

import (
	"time"

	"github.com/example/monorepo/internal/auth/domain"
)

// JwtService handles JWT token generation and validation.
type JwtService struct {
	userRepo  domain.UserRepository
	tokenRepo domain.TokenRepository
}

// NewJwtService creates a new JWT service.
func NewJwtService(userRepo domain.UserRepository, tokenRepo domain.TokenRepository) *JwtService {
	return &JwtService{userRepo: userRepo, tokenRepo: tokenRepo}
}

// GenerateToken generates a JWT token for the given user.
func (s *JwtService) GenerateToken(user *domain.User) (string, error) {
	token := &domain.Token{
		Value:     "fake-jwt-token-for-" + user.ID,
		UserID:    user.ID,
		Role:      string(user.Role),
		ExpiresAt: time.Now().Add(time.Hour),
	}
	if err := s.tokenRepo.Save(token); err != nil {
		return "", err
	}
	return token.Value, nil
}

// ValidateToken validates a JWT token and returns the user.
func (s *JwtService) ValidateToken(tokenValue string) (*domain.User, error) {
	token, err := s.tokenRepo.FindByValue(tokenValue)
	if err != nil || token == nil {
		return nil, domain.ErrUnauthorized
	}
	user, err := s.userRepo.FindByID(token.UserID)
	if err != nil {
		return nil, domain.ErrNotFound
	}
	return user, nil
}

// RequireAuth requires a valid authorization header and returns the user.
func (s *JwtService) RequireAuth(tokenValue string) (*domain.User, error) {
	if tokenValue == "" {
		return nil, domain.ErrUnauthorized
	}
	user, err := s.ValidateToken(tokenValue)
	if err != nil {
		return nil, err
	}
	return user, nil
}

File: examples/rust/src/auth/service/jwt.rs
use crate::auth::domain::{Token, TokenRepository, User, UserRepository};
use crate::auth::domain::error::AuthError;
use std::time::{Duration, SystemTime};

pub struct JwtService {
    user_repo: Box<dyn UserRepository>,
    token_repo: Box<dyn TokenRepository>,
}

impl JwtService {
    pub fn new(user_repo: Box<dyn UserRepository>, token_repo: Box<dyn TokenRepository>) -> Self {
        Self { user_repo, token_repo }
    }

    pub fn generate_token(&self, user: &User) -> Result<String, AuthError> {
        let token = Token {
            value: format!("fake-jwt-token-for-{}", user.base.id),
            user_id: user.base.id.clone(),
            role: format!("{:?}", user.role),
            expires_at: SystemTime::now() + Duration::from_secs(3600),
        };
        self.token_repo.save(&token)?;
        Ok(token.value)
    }

    pub fn validate_token(&self, token_value: &str) -> Result<Option<User>, AuthError> {
        let token = self.token_repo.find_by_value(token_value)?;
        let token = match token {
            Some(t) => t,
            None => return Ok(None),
        };
        let user = self.user_repo.find_by_id(&token.user_id)?;
        Ok(user)
    }

    pub fn require_auth(&self, token_value: Option<&str>) -> Result<User, AuthError> {
        let token_value = token_value.ok_or_else(AuthError::unauthorized)?;
        let user = self.validate_token(token_value)?
            .ok_or_else(|| AuthError::not_found("token"))?;
        Ok(user)
    }
}
